src/controllers/userController.ts

```diff
@@ -0,0 +1,139 @@
+import { Request, Response } from "express";
+import { CreateUserDTO, EditUserDTO } from "../types";
+import {
+  saveUser,
+  fetchUserByEmail,
+  isValidEmail,
+  usernameExists,
+  generatePassword,
+  getUserById,
+  updateUser,
+} from "../models/user";
+
+import {
+  EMAIL_TAKEN_ERROR,
+  successResponse,
+  USERNAME_TAKEN_ERROR,
+  VALIDATION_ERROR,
+  INTERNAL_SERVER_ERROR,
+  httpcode,
+  USER_NOT_FOUND_ERROR,
+} from "../models/response";
+
+export const createUser = async (
+  req: Request<{}, {}, CreateUserDTO>,
+  res: Response
+) => {
+  try {
+    const { email, lastName, firstName, username, password } = req.body;
+
+    if (!email || !lastName || !firstName || !username) {
+      return res.status(httpcode.BAD_REQUEST).json(VALIDATION_ERROR);
+    }
+
+    const validEmail = isValidEmail(email);
+    if (!validEmail) {
+      return res.status(httpcode.BAD_REQUEST).json(VALIDATION_ERROR);
+    }
+
+    const userWithEmailExists = await fetchUserByEmail({ email });
```

Comment:
I did something similar but noticed another project using Prisma errors for this case, which could be more efficient as it avoids multiple DB checks.

Reply:
Yeah I saw something quite similar in a PR I reviewed recently, I thought it was pretty nifty.